lib/Sema/TypeCheckType.cpp

```diff
@@ -1304,9 +1304,18 @@ static SelfTypeKind getSelfTypeKind(DeclContext *dc,
 
   // In enums and structs, 'Self' is just a shorthand for the nominal type,
   // and can be used anywhere.
-  if (!typeDC->getSelfClassDecl())
+  auto *const classDecl = typeDC->getSelfClassDecl();
+  if (!classDecl)
     return SelfTypeKind::StaticSelf;
 
+  // In a final class, 'Self' is a shorthand for the nominal type.
+  if (classDecl->isFinal()) {
+    // FIXME: Extending this to the pre-stability accepted covariant 'Self'
+    // returns is an ABI break due to mangling differences.
```

Comment:
Perhaps a better approach is to still resolve Self as DynamicSelfType in final classes, but make the type of `self` into a non-DynamicSelfType. See `computeSelfParam()` in ASTContext.cpp.

You'll need to mess with the constraint solver to make this work, though, hopefully it's not too difficult:

```
final class C {
  func f() -> Self { return self }
}
```

Also, you can probably continue to resolve Self to the static type inside a stored property initializer -- that is, if `dc` (or one of its parents) is a `PatternBindingInitializer`.

Reply:
Are you suggesting that we resolve invariant and contravariant references as DynamicSelfType too?

Reply:
I think those could be resolved to the static Self type in a final class (and diagnosed in non-final as before).